Rework Book (#1230)
graphql-rust/juniper: book/src/types/objects/context.md

Context
=======

_Context_ is a feature in [Juniper] that lets [field][4] resolvers access global data, most commonly database connections or authentication information.

Let's say that we have a simple `User`s database in a `HashMap`:
```rust
# #![allow(dead_code)]
# use std::collections::HashMap;
#
struct Database {
    users: HashMap<i32, User>,
}

struct User {
    id: i32,
    name: String,
    friend_ids: Vec<i32>,
}
#
# fn main() {}
```

> We would like a `friends` [field][4] on `User` that returns a list of `User` [objects][4]. In order to write such a [field][4] though, the `Database` must be queried. To solve this, we mark the `Database` as a valid context type and assign it to the `User` [object][4]. To gain access to the context, we need to specify an argument with the same type as the specified `context` for the type:

Reviewer: ```suggestion
We would like to define a `friends` [field][4] on `User` that returns a list of `User` [objects][4]. In order to write such a [field][4] we need to query a `Database`. To accomplish this we must first mark the `Database` as a valid context type and then assign it to the `User` [object][4]. To gain access to the context in the `friends` [field][4], we need to specify an argument with the same type as the specified context:
```

impl User {
    // Inject the `Database` context by specifying an argument with the 
    // context type: 
    // - the type must be a reference;
    // - the name of the argument SHOULD be `context` (or `ctx`).
    fn friends<'db>(&self, context: &'db Database) -> Vec<&'db User> {
        //                 ^^^^^^^ or `ctx`, up to your preference
        self.friend_ids.iter()
            .map(|id| {
                context.users.get(&id).expect("could not find `User` with ID")
            })
            .collect()
    }

    fn friend<'db>(
        &self, 
        id: i32, 
        // Alternatively, the context argument may be marked with an attribute,
        // and thus, named arbitrary.
        #[graphql(context)] db: &'db Database,
        //        ^^^^^^^ or `ctx`, up to your preference
    ) -> Option<&'db User> {
        self.friend_ids.contains(&id).then(|| {
            db.users.get(&id).expect("could not find `User` with ID")
        })
    }

    fn name(&self) -> &str { 
        self.name.as_str() 
    }

    fn id(&self) -> i32 { 
        self.id 
    }
}
#
# fn main() {}
```

### Mutating and mutable references

> Context cannot be specified by a mutable reference, because concurrent [fields][4] resolving may be performed. If there is something in a context that requires access by a mutable reference, the context type should follow the [_interior mutability_ pattern][5] for that (e.g. using `RwLock`, `RefCell` or similar).

Reviewer: ```suggestion
Context cannot be a mutable reference as [fields][4] may be resolved concurrently. If something in the context requires a mutable reference, the context type should leverage the [_interior mutability_ pattern][5] (e.g. use `RwLock`, `RefCell` or similar).
```
